src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets

```diff
@@ -637,10 +599,41 @@ robocopy /S /NP %EXECUTION_DIR%int\%(Identity)\ %EXECUTION_DIR%native\
 
   <Target Name="SetupTestProperties" DependsOnTargets="CheckTestPlatforms;CheckTestCategories" />
 
+  <PropertyGroup>
+    <LaunchSettingsOutputFileFullPath>$(MSBuildProjectDirectory)/$(AppDesignerFolder)/launchSettings.json</LaunchSettingsOutputFileFullPath>
+    <LaunchSettingsInputFileFullPath>$(MSBuildThisFileDirectory)launchSettings.json</LaunchSettingsInputFileFullPath>
+  </PropertyGroup>
+
+  <!--
+    Invoked either manually by the build.proj to generate or update the launch settings file
+    or by the build if it doesn't exist yet.
+  -->
+  <Target Name="GenerateLaunchSettingsFile" Condition="$(IsInvokableTestProject) == 'true' AND '$(GenerateLaunchSettings)' == 'true'">
```

Comment:
You should have an input/output here to avoid copying this every time.

Reply:
Where is GenerateLaunchSettings expected to get set?

Reply:
> You should have an input/output here to avoid copying this every time.

That target is only hooked to build if the file doesn't already exist: https://github.com/dotnet/buildtools/pull/2168/files/d2d2905092f1931b1c35614addecd5bbf52d0a66#diff-425cb0dcf83ca9680e4a736c021dabe8R619. This is only to support the scenario when someone adds a new project or accidentally deleted his launchSettings.json file.

The strategy changed from generating the launchSettings.json during the test project build to hooking it into the repo build: https://github.com/dotnet/corefx/pull/32380/files#diff-3ad318a39efd856540e8cf813f20ec09R53. I talked with @joperezr about that and he also recommended that approach. This solves the race as we only once create the files when calling build and not for each project for each configuration (which caused the race in the first place).

> Where is GenerateLaunchSettings expected to get set?

It's set in the Directory.Build.props in corefx with my accompanying PR. I decided to make it an opt-in as the coreclr tests are also using the tests.targets somehow and I don't want to drop random files there.

Reply:
OK I think I can accept although it will require a full build for anyone adding new test projects but I suspect we can live with that. I would put all the conditions in one or the other place and not have them split between the target and the PropertyGroup that will make it clearer as to what is controlling it.

Reply:
ok cool. just one question, why do you think it requires a full build when adding a new test project? the target is hooked into the test assembly build if the file doesn't yet exist. or do you mean with full build the the test assemblies build and not the root build?

Reply:
> I would put all the conditions in one or the other place and not have them split between the target and the PropertyGroup that will make it clearer as to what is controlling it.

The condition on the PropertyGroup is only for the scenario after the build, if the file doesn't already exist to create it --> new project added. The conditions on the target also apply here.

The condition on the target applies to both the build and to the manual invocation from corefx. We don't want the file check here as it should overwrite already existing files when a new buildtools version comes in --> when building from root.

Reply:
Unless I missed something you currently only plan to run this target once up front and not during the test assembly build any longer, or did I miss-understand you on that? If it runs during the build as well as upfront then I agree it will work for both scenarios.

Reply:
>The condition on the target applies to both the build and to the manual invocation from corefx. We don't want the file check here as it should overwrite already existing files when a new buildtools version comes in --> when building from root.

That is not very clear at all based on how things are setup. If that is the goal then we should add a comment to that affect.

Reply:
right, we also run during the build but only if the file doesn't already exist (i.e. new project or the file was deleted).

Reply:
ok will do.